### ISSUE TITLE
cattle agent with PXE / Stateless hosts

### ISSUE THREAD

**Issue by justinrush** [rancher/os#2880]
**RancherOS Version: (v1.5.3)**

**Where are you running RancherOS? (baremetal)**

We use RKE to deploy our clusters for rancher.  We are adding some new baremetal stateless servers into the cluster but are unsure how to get them to join the cluster and stay in it.  I'm sure if I RKE up on my modified RKE file, the hosts will get added into the cluster, however, when they come up back, they won't have the cattle agent docker image on there and it won't re-join the cluster.  I found some old docs for Rancher 1.2 (https://rancher.com/docs/os/v1.x/en/about/running-rancher-on-rancheros/), but the API doesn't appear to exist in rancher v2.2.6.  Also, is the registration key the cattle-credentials/token secret value?

**Comment by niusmallnan** [comment]
I am not clear your points.
Can you show me more steps so that I can reproduce your scene?

**Comment by justinrush** [comment]
Sure, the way we have deployed k8s nodes in the past:
1. provision VM
2. run puppet with the k8s role
3. update RKE config file to add node to cluster
4. run rke up --update-only to get the node added to the kubernetes cluster.  This pulls down the cattle agent and various other containers needed in the cluster and sets them to always restart.
5. reboot VM, cattle-agent container starts on boot as it is persisted on disk, then it automatically rejoins the cluster.

The way we're trying to go with baremetal nodes and racheros / stateless:
1. rack and stack new hosts, assign proper vlans, etc.
2. add DHCP reservation for new host
3. Boot host, the host will reach out to the PXE TFTP server and get the cloud config and be ready to go
4. update RKE config file to add node to cluster
5. run rke up --update-only to get the node added to the kubernetes cluster.  Like the above, the node is full joined and usable in the the cluster.
6. Reboot node.  Node comes up with no containers because there is no persistent storage

**Comment by niusmallnan** [comment]
https://rancher.com/docs/os/v1.x/en/installation/running-rancheros/server/pxe/
You can add `rancher.state.dev=LABEL=RANCHER_STATE rancher.state.autoformat=[/dev/sda] rancher.state.wait` to cmdline when booting by iPXE, it can autoformat the disk.

**Comment by justinrush** [comment]
Yes, i know we can do that, but I want them to be stateless with no disks.

**Comment by niusmallnan** [comment]
@justinrush You can try to add the registration command to `/etc/rc.local`, such as:
```
#cloud-config
rancher:
write_files:
  - path: /etc/rc.local
    permissions: "0755"
    owner: root
    content: |
      #!/bin/bash
      wait-for-docker
      # rancher2.x registration commands here
```

https://rancher.com/docs/os/v1.x/en/installation/configuration/running-commands/#running-docker-commands

**Comment by niusmallnan** [comment]
Very interesting scene, please think about the solution when you have a chance. @Jason-ZW @kingsd041

**Comment by xeor** [comment]
Anyone came up with a solution to this?
